editor/language_server/src/lsp.rs

```diff
@@ -0,0 +1,139 @@
+        //      - There is a tradeoff between evaluating everything at once on change and lazily
+        //        evaluating on queries. For now, we need to do it in this step because some
+        //        diagnostics (i.e. validation) cannot be done without performing evaluation
+        //        anyway.
+        //        A future flow could be to use a "query" model, whereby we incrementally parse,
+        //        evaluate, and publish diagnostics localized to a single statement.
+        //        But we are far away from that being important.
+        //   3. Since we're already here, publish any diagnostics we discovered.
+        //
+        // We cache both the original program AST and evaluated AST so we can answer later queries
+        // for original/optimized statements without re-evaluation.
+        
+        let context = ProgramContext::default().lint(true);
+        let ScanResult { tokens, diagnostics: scan_diags } = scan(&*text);
+        let ParseResult { program, diagnostics: parse_diags } = parse_statements(tokens, &text);
+        let lint_diags = lint_stmt(&program, &text);
+        let EvaluationResult { simplified, diagnostics: eval_diags } = evaluate(program.clone(), &context).expect("Evaluation failed.");
+
+        self.doc_registry().get_mut().insert(doc.clone(), ProgramInfo { original: program, simplified });
+
+        let diags = [scan_diags, parse_diags, lint_diags, eval_diags]
+            .iter()
+            .flat_map(|diags| convert_diagnostics(diags, "slide", &doc, &text))
+            .collect();
+
+        self.client.publish_diagnostics(doc.clone(), diags, version).await;
+    }
+
+    fn close(&self, doc: &Url) {
+        self.doc_registry().get_mut().remove(doc);
+    }
+}
+
+#[tower_lsp::async_trait]
+impl LanguageServer for SlideLS {
+    async fn initialize(&self, _params: InitializeParams) -> Result<InitializeResult> {
+        // let hover_provider = Some(HoverProviderCapability::Simple(true));
+        let text_document_sync = Some(TextDocumentSyncCapability::Options(TextDocumentSyncOptions {
+            open_close: Some(true),
+            change: Some(TextDocumentSyncKind::Full),
+            ..TextDocumentSyncOptions::default()
+        }));
+        Ok(InitializeResult{
+            capabilities: ServerCapabilities {
+                text_document_sync,
+                // hover_provider,
+                ..ServerCapabilities::default()
+            },
+            ..InitializeResult::default()
+        })
+    }
+
+    async fn initialized(&self, _params: InitializedParams) {
+        self.client
+            .log_message(MessageType::Info, "Slide language server initialized.")
+            .await;
+    }
+
+    async fn shutdown(&self) -> Result<()> {
+        Ok(())
+    }
+
+    async fn did_open(&self, params: DidOpenTextDocumentParams) {
+        let TextDocumentItem { uri, text, version, .. } = params.text_document;
+        self.change(uri, text, Some(version)).await;
+    }
```

Comment:
shouldnt this return something just or i guess its async so maybe not.

Reply:
yeah nothing to ret